third_party/amd/lib/TritonAMDGPUToLLVM/OptimizeLDSUsage.cpp

```diff
@@ -0,0 +1,305 @@
+        src.getMDim(), src.getNDim(), src.getIsTransposed(),
+        src.getCTALayout());
+  if (auto src = layout.dyn_cast<triton::gpu::AMDWmmaEncodingAttr>())
+    return triton::gpu::AMDWmmaEncodingAttr::get(ctx, warpsPerCTA,
+                                                 src.getCTALayout());
+  if (auto src = layout.dyn_cast<triton::gpu::BlockedEncodingAttr>())
+    return triton::gpu::BlockedEncodingAttr::get(
+        ctx, src.getSizePerThread(), src.getThreadsPerWarp(), warpsPerCTA,
+        src.getOrder(), src.getCTALayout());
+  if (auto src = layout.dyn_cast<triton::gpu::DotOperandEncodingAttr>()) {
+    return triton::gpu::DotOperandEncodingAttr::get(
+        ctx, src.getOpIdx(), createTmpLayout(src.getParent(), warpsPerCTA),
+        src.getKWidth());
+  }
+  if (auto src = layout.dyn_cast<triton::gpu::SliceEncodingAttr>())
+    return triton::gpu::SliceEncodingAttr::get(
+        ctx, src.getDim(), createTmpLayout(src.getParent(), warpsPerCTA));
+  assert("Encountered unsupported layout");
+  return Attribute();
+}
+
+/**
+ * Creates two chained convert layout operations
+ *
+ * %1 = cvtOp %0 (srcLayout -> dstLayout)
+ * ->
+ * %1 = cvtOp %0 (srcLayout -> dstLayout)
+ * %2 = cvtOp %0 (srcLayout -> tmpLayout)
+ * %3 = cvtOp %1 (tmpLayout -> dstLayout)
+ *
+ * @param builder
+ * @param cvtOp original operation
+ * @param tmpLayout
+ * @return pair of created operations
+ */
+static std::pair<triton::gpu::ConvertLayoutOp, triton::gpu::ConvertLayoutOp>
+createNewConvertOps(OpBuilder &builder, triton::gpu::ConvertLayoutOp &cvtOp,
+                    Attribute tmpLayout) {
+  auto srcType = cvtOp.getSrc().getType();
+  auto dstType = cvtOp.getType();
+
+  auto newDstType = RankedTensorType::get(
+      dstType.getShape(), dstType.getElementType(), dstType.getEncoding());
+  RankedTensorType newSrcType = RankedTensorType::get(
+      srcType.getShape(), srcType.getElementType(), tmpLayout);
+
+  auto tmpCvt = builder.create<triton::gpu::ConvertLayoutOp>(
+      cvtOp.getLoc(), newSrcType, cvtOp.getSrc());
+  auto newEpilogueCvt = builder.create<triton::gpu::ConvertLayoutOp>(
+      cvtOp.getLoc(), newDstType, tmpCvt);
+
+  return std::make_pair(tmpCvt, newEpilogueCvt);
+}
+
+// Try to reduce LDS usage of cvt op.
+//
+// Consider mfma->blocked conversion as an example.
+// LDS reduction is possible by changing the shape of WarpsPerCta attribute in
+// mfma layout. The implicit LDS usage of cvt(mfma->blocked) op depends on the
+// number of warps per CTA that mfma layout uses along x dimension and block
+// layout uses across y dimension.
+//
+// clang-format off
+//
+// LDS usage of this op is roughly calculated as:
+// LDS_USAGE = getShapePerCTA(mfma_layout)[0] * getShapePerCTA(blocked_layout)[1] * sizeof(data_type)
+// LDS_USAGE = warpsPerCTA(mfma_layout)[0] * warpsPerCta(blocked_layout)[1] * C,
+// where C = 32 * sizePerWarp(blocked_layout)[1] * threadsPerWarp(blocked_layout)[1] * sizeof(data_type)
```

Comment:
Why is 32 hardcoded? Is it assuming mfma32 is used?

Reply:
To be honest, I did not look deep into this comment, just copied it from original algorithm.
It was implemented a log ago, we probably had only mfma32 at the time.

I'll take a closer look and adjust.